src/neo/SmartContract/Native/NotaryContract.cs

```diff
@@ -0,0 +1,322 @@
+        }
+
+        /// <summary>
+        /// GetNotaryNodes returns public keys of notary nodes.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <returns></returns>
+        private ECPoint[] GetNotaryNodes(DataCache snapshot)
+        {
+            return RoleManagement.GetDesignatedByRole(snapshot, Role.Notary, Ledger.CurrentIndex(snapshot));
+        }
+
+        /// <summary>
+        /// GetMaxNotValidBeforeDelta is Notary contract method and returns the maximum NotValidBefore delta.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <returns>NotValidBefore</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public uint GetMaxNotValidBeforeDelta(DataCache snapshot)
+        {
+            return (uint)(BigInteger)snapshot[CreateStorageKey(PreMaxNotValidBeforeDelta)];
+        }
+
+        /// <summary>
+        /// SetMaxNotValidBeforeDelta is Notary contract method and sets the maximum NotValidBefore delta.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="value">Value</param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private void SetMaxNotValidBeforeDelta(ApplicationEngine engine, uint value)
+        {
+            if (value > engine.ProtocolSettings.MaxValidUntilBlockIncrement / 2 || value < ProtocolSettings.Default.ValidatorsCount) throw new FormatException(string.Format("MaxNotValidBeforeDelta cannot be more than {0} or less than {1}", engine.ProtocolSettings.MaxValidUntilBlockIncrement / 2, ProtocolSettings.Default.ValidatorsCount));
+            if (!CheckCommittee(engine)) throw new InvalidOperationException();
+            engine.Snapshot.GetAndChange(CreateStorageKey(PreMaxNotValidBeforeDelta)).Set(value);
+        }
+
+        /// <summary>
+        /// GetNotaryServiceFeePerKey is Notary contract method and returns the NotaryServiceFeePerKey delta.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <returns>NotaryServiceFeePerKey</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public long GetNotaryServiceFeePerKey(DataCache snapshot)
+        {
+            return (long)(BigInteger)snapshot[CreateStorageKey(PreNotaryServiceFeePerKey)];
+        }
+
+        /// <summary>
+        /// SetNotaryServiceFeePerKey is Notary contract method and sets the NotaryServiceFeePerKey delta.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="value">value</param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private void SetNotaryServiceFeePerKey(ApplicationEngine engine, long value)
+        {
+            if (value < 0) throw new FormatException("NotaryServiceFeePerKey cannot be less than 0");
```

Comment:
Let's add the maximum constraint on `value`. I can suggest using 1_0000_0000 GAS, although even that is too expensive.

We may tune this value in future.